diff --git a/CMakeLists.txt b/CMakeLists.txt
index 850fd7e..48e08e5 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -16,8 +16,10 @@ set(FETCHCONTENT_QUIET OFF)
 
 
 # Debug
+add_compile_options(-Wno-gnu-alignof-expression)
 add_compile_definitions($<$<CONFIG:Debug>:_DBG> $<$<CONFIG:RelWithDebInfo>:_DBG>)
 add_compile_definitions(_mochi_debug_ownership)
+add_compile_definitions(_mochi_debug_validator)
 
 
 # Sanitizer
diff --git a/mochi/asset/mesh.cc b/mochi/asset/mesh.cc
index 9d4b978..429c619 100644
--- a/mochi/asset/mesh.cc
+++ b/mochi/asset/mesh.cc
@@ -35,7 +35,7 @@
 namespace mochi::asset
 {
   
-  auto vertex_i = rhi::info<rhi::buffer>::make(
+  rhi::info<rhi::buffer> vertex_i(
     sizeof(vertex_t),
     rhi::vt::make_list<
       vec3<f32>, // Position
@@ -191,11 +191,11 @@ namespace mochi::asset
     m_data = rhi::buffer::make(
       core.sub<module::device>(), core.sub<module::memory>(),
       vertex_i, final_data.size(),
-      flags(rhi::BufferUsage::VertexBuffer) | rhi::BufferUsage::TransferDst,
-      flags(rhi::BufferCreate::HostSequentialWrite) | rhi::BufferCreate::Mapped,
+      flags(rhi::BufferUsage::DeviceAddress) | rhi::BufferUsage::TransferDst,
+      rhi::BufferCreate::HostSequentialWrite,
       rhi::BufferLocation::PreferDevice,
       [&final_data](void* _data) {
-        memcpy(_data, final_data.data(), vertex_i->stride() * final_data.size());
+        memcpy(_data, final_data.data(), vertex_i.stride() * final_data.size());
       }
     );
     m_offs = std::move(final_offs);
diff --git a/mochi/asset/mesh.hh b/mochi/asset/mesh.hh
index 19d4697..f1e4d11 100644
--- a/mochi/asset/mesh.hh
+++ b/mochi/asset/mesh.hh
@@ -34,7 +34,7 @@ namespace mochi::asset
   };
   
   /** @brief Buffer info describing the vertex_t memory layout. */
-  extern sptr<rhi::info<rhi::buffer>> vertex_i;
+  extern rhi::info<rhi::buffer> vertex_i;
 
 
   
diff --git a/mochi/core.cc b/mochi/core.cc
index cfbef84..9c4a06f 100644
--- a/mochi/core.cc
+++ b/mochi/core.cc
@@ -10,7 +10,6 @@
 */
 
 
-#include "mochi/asset/texture.hh"
 #include "mochi/basis.hh"
 #include "mochi/core.hh"
 #include "mochi/ecs/camera.hh"
@@ -18,7 +17,9 @@
 #include "mochi/ecs/camera.hh"
 #include "mochi/ecs/point_light.hh"
 #include "mochi/ecs/transform.hh"
+#include "mochi/rhi/cmd.hh"
 #include "mochi/rhi/image.hh"
+#include "mochi/rhi/listPush.hh"
 #include "mochi/rhi/pipeline.hh"
 #include "mochi/rhi/buffer.hh"
 #include "mochi/module/device.hh"
@@ -26,6 +27,8 @@
 #include "mochi/module/display.hh"
 #include "mochi/module/memory.hh"
 #include "mochi/module/renderer.hh"
+#include "mochi/rhi/slotPush.hh"
+#include "vulkan/vulkan.hpp"
 #include <chrono>
 #include <vulkan/vulkan_raii.hpp>
 
@@ -91,7 +94,7 @@ namespace mochi
   }
 
 
-  fun core::paint(vk::raii::CommandBuffer &cmd, rhi::render_target &target) -> void
+  fun core::paint(rhi::cmd &cmd, rhi::render_target &target) -> void
   {
     auto &mem = sub<module::memory>();
 
@@ -133,8 +136,8 @@ namespace mochi
 
 
     auto extent = sub<module::display>().extent(); 
-    cmd.setViewport(0, {vk::Viewport(0.0f, 0.0f, (float)extent.width, (float)extent.height, 0.0f, 1.0f)});
-    cmd.setScissor(0, {vk::Rect2D({0, 0}, extent)});
+    cmd.get().setViewport(0, {vk::Viewport(0.0f, 0.0f, (float)extent.width, (float)extent.height, 0.0f, 1.0f)});
+    cmd.get().setScissor(0, {vk::Rect2D({0, 0}, extent)});
 
     
 
@@ -153,15 +156,8 @@ namespace mochi
       if (!renderable.mesh)
         continue;
 
-      
       auto mesh = renderable.mesh;
 
-      if (mesh->data().get() != last_buffer) {
-        cmd.bindVertexBuffers(0, {mesh->data()->get()}, {0}); 
-        last_buffer = mesh->data().get();
-      }
-
-
       
       u32 i{};
       for (auto &subsur: mesh->offs())
@@ -170,11 +166,11 @@ namespace mochi
         auto desc = material->desc(target);
 
         if (desc->pipeline.get() != last_pipeline) {
-          cmd.bindPipeline(vk::PipelineBindPoint::eGraphics, *desc->pipeline->get());
+          cmd.setPipeline(desc->pipeline.get());
           last_pipeline = desc->pipeline.get();
         }
 
-        cmd.bindDescriptorSets(
+        cmd.get().bindDescriptorSets(
           vk::PipelineBindPoint::eGraphics,
           *desc->pipeline->layout(),
           0,
@@ -184,7 +180,7 @@ namespace mochi
 
 
         if (material->is_texture()) {
-          cmd.bindDescriptorSets(
+          cmd.get().bindDescriptorSets(
             vk::PipelineBindPoint::eGraphics,
             *desc->pipeline->layout(),
             1,
@@ -193,16 +189,10 @@ namespace mochi
           );
         }
 
+        cmd.writePushConstant(rhi::listPush(std::tuple{transform.model, mesh->data()->address()}));
+        
 
-        auto model_mat = transform.model;
-        cmd.pushConstants<mochi::mat4<f32>>(
-          *last_pipeline->layout(),
-          vk::ShaderStageFlagBits::eVertex | vk::ShaderStageFlagBits::eFragment,
-          0,
-          model_mat
-        );
-
-        cmd.draw(subsur.size(), 1, subsur.off(), 0);
+        cmd.draw({subsur.off(), subsur.size()}, {0, 1});
         i++;
       }
 
@@ -231,7 +221,9 @@ namespace mochi
 
     ren.begin_pass(cmd, target, {0.1f, 0.1f, 0.1f, 1.0f});
 
-    /* paint */ paint(cmd, target);
+    rhi::cmd mochi_cmd(&cmd);
+
+    /* paint */ paint(mochi_cmd, target);
 
     ren.end_pass(cmd, target);
 
diff --git a/mochi/core.hh b/mochi/core.hh
index 7a015f8..e3b6bca 100644
--- a/mochi/core.hh
+++ b/mochi/core.hh
@@ -23,6 +23,7 @@
 #include <functional>
 #include <tuple>
 #include <vulkan/vulkan_raii.hpp>
+#include "mochi/rhi/cmd.hh"
 #include "src/entt/entt.hpp"
 
 
@@ -88,7 +89,7 @@ namespace mochi
       fun run() -> void;
 
     private:
-      fun paint(vk::raii::CommandBuffer &cmd, rhi::render_target &target) -> void;
+      fun paint(rhi::cmd &cmd, rhi::render_target &target) -> void;
       fun draw() -> void;
   };
 
diff --git a/mochi/ecs/camera.cc b/mochi/ecs/camera.cc
index 7decc5d..b0882ed 100644
--- a/mochi/ecs/camera.cc
+++ b/mochi/ecs/camera.cc
@@ -23,7 +23,7 @@
 namespace mochi::ecs
 {
 
-  auto camera_i = rhi::info<rhi::buffer>::make(
+  rhi::info<rhi::buffer> camera_i(
     sizeof(camera_t),
     rhi::vt::make_list<
       mat4<f32>, // View
diff --git a/mochi/ecs/camera.hh b/mochi/ecs/camera.hh
index 949f99c..fb37d93 100644
--- a/mochi/ecs/camera.hh
+++ b/mochi/ecs/camera.hh
@@ -30,7 +30,7 @@ namespace mochi::ecs
   };
   
   /** @brief Buffer info describing the camera3d_t memory layout. */
-  extern sptr<rhi::info<rhi::buffer>> camera_i;
+  extern rhi::info<rhi::buffer> camera_i;
 
 
 
diff --git a/mochi/ecs/point_light.cc b/mochi/ecs/point_light.cc
index d5b3c87..f0d2709 100644
--- a/mochi/ecs/point_light.cc
+++ b/mochi/ecs/point_light.cc
@@ -23,7 +23,7 @@
 namespace mochi::ecs
 {
 
-  auto point_light_i = rhi::info<rhi::buffer>::make(
+  rhi::info<rhi::buffer> point_light_i(
     sizeof(point_light_t),
     rhi::vt::make_list<
       vec4<f32>, // Position
diff --git a/mochi/ecs/point_light.hh b/mochi/ecs/point_light.hh
index e0aaa8c..2f94e5c 100644
--- a/mochi/ecs/point_light.hh
+++ b/mochi/ecs/point_light.hh
@@ -30,7 +30,7 @@ namespace mochi::ecs
   };
   
   /** @brief Buffer info describing the light3d_t memory layout. */
-  extern sptr<rhi::info<rhi::buffer>> point_light_i;
+  extern rhi::info<rhi::buffer> point_light_i;
 
 
 
diff --git a/mochi/module/device.cc b/mochi/module/device.cc
index 55eaa3e..5984ebd 100644
--- a/mochi/module/device.cc
+++ b/mochi/module/device.cc
@@ -46,10 +46,14 @@ namespace mochi::module
 
     vk::PhysicalDeviceFeatures features;
     features.samplerAnisotropy = VK_TRUE;
+    features.fillModeNonSolid = VK_TRUE;
 
+    vk::PhysicalDeviceVulkan12Features features12{};
+    features12.bufferDeviceAddress = VK_TRUE;
     
     vk::PhysicalDeviceDynamicRenderingFeatures dynamicRenderingFeature;
     dynamicRenderingFeature.dynamicRendering = VK_TRUE;
+    dynamicRenderingFeature.pNext = features12;
 
 
     vk::DeviceCreateInfo dev_info({}, queue_infos, {}, extensions, &features, &dynamicRenderingFeature);
@@ -97,11 +101,11 @@ namespace mochi::module
     vk::CommandPoolCreateInfo main_pool_info(vk::CommandPoolCreateFlagBits::eResetCommandBuffer, main_q().family());
     m_mainPool = vk::raii::CommandPool(vk_device, main_pool_info);
 
-    vk::CommandPoolCreateInfo transfer_pool_info(vk::CommandPoolCreateFlagBits::eTransient, transfer_q().best().family());
+    vk::CommandPoolCreateInfo transfer_pool_info(vk::CommandPoolCreateFlagBits::eResetCommandBuffer, transfer_q().best().family());
     m_transferPool = vk::raii::CommandPool(vk_device, transfer_pool_info);
 
 
-    vk::CommandBufferAllocateInfo alloc_info(*m_mainPool, vk::CommandBufferLevel::ePrimary, 1);
+    vk::CommandBufferAllocateInfo alloc_info(*m_transferPool, vk::CommandBufferLevel::ePrimary, 1);
     m_transferBuf = std::move(vk::raii::CommandBuffers(vk_device, alloc_info).front());
   }
 
diff --git a/mochi/module/memory.cc b/mochi/module/memory.cc
index 24cbbb2..2e22b56 100644
--- a/mochi/module/memory.cc
+++ b/mochi/module/memory.cc
@@ -68,10 +68,11 @@ namespace mochi::module
 
 
     VmaAllocatorCreateInfo allocatorInfo = {};
+    allocatorInfo.vulkanApiVersion = VK_API_VERSION_1_4;
     allocatorInfo.physicalDevice = *device.phys_dev();
     allocatorInfo.device = *device.vdevice();
     allocatorInfo.instance = *bridge.inst();
-    allocatorInfo.vulkanApiVersion = VK_API_VERSION_1_4;
+    allocatorInfo.flags = VMA_ALLOCATOR_CREATE_BUFFER_DEVICE_ADDRESS_BIT;
 
     vmaCreateAllocator(&allocatorInfo, &m_allocator);
   }
@@ -88,7 +89,7 @@ namespace mochi::module
 
   fun memory::camera_ubo(u64 required_count) -> sptr<rhi::buffer>
   {
-    if (!m_camera_ubo || (m_camera_ubo->size() / ecs::camera_i->stride()) < required_count)
+    if (!m_camera_ubo || (m_camera_ubo->size() / ecs::camera_i.stride()) < required_count)
       m_camera_ubo = rhi::buffer::make(
         m_device, *this,
         ecs::camera_i, std::max<u64>(10, required_count),
@@ -101,7 +102,7 @@ namespace mochi::module
 
   fun memory::light_ubo(u64 required_count) -> sptr<rhi::buffer>
   {
-    if (!m_light_ubo || (m_light_ubo->size() / ecs::point_light_i->stride()) < required_count)
+    if (!m_light_ubo || (m_light_ubo->size() / ecs::point_light_i.stride()) < required_count)
       m_light_ubo = rhi::buffer::make(
         m_device, *this,
         ecs::point_light_i, std::max<u64>(100, required_count),
diff --git a/mochi/module/resource.cc b/mochi/module/resource.cc
index c5983e0..8564670 100644
--- a/mochi/module/resource.cc
+++ b/mochi/module/resource.cc
@@ -12,9 +12,9 @@
 
 #include "mochi/basis.hh"
 #include "mochi/except.hh"
+#include "mochi/types.hh"
 #include "mochi/rhi/rhi.hh"
 #include "mochi/rhi/vtype.hh"
-#include "mochi/types.hh"
 #include "mochi/asset/mesh.hh"
 #include "mochi/asset/texture.hh"
 #include "mochi/ecs/camera.hh"
@@ -22,10 +22,10 @@
 #include "mochi/rhi/convert.hh"
 #include "mochi/rhi/pipeline.hh"
 #include "mochi/rhi/render_target.hh"
-#include "mochi/rhi/descset.hh"
+#include "mochi/rhi/listPush.hh"
+#include "mochi/rhi/listDesc.hh"
 #include "mochi/rhi/slotPush.hh"
 #include "mochi/rhi/slotDesc.hh"
-#include "mochi/rhi/slotVertex.hh"
 #include "mochi/module/resource.hh"
 #include "mochi/vfs/vfs.hh"
 #include "vulkan/vulkan.hpp"
@@ -45,25 +45,31 @@ namespace mochi::module
   resource::resource(module::device &device, module::memory &memory)
     : m_device(device)
     , m_memory(memory)
+    , shader_color_i(nil)
+    , shader_texture_i(nil)
   {
     create_pool();
 
 
-    auto Transform = rhi::info<rhi::slotPush>::make(rhi::vt::make<mat4<f32>>(), flags(rhi::ShaderStage::Vertex) | rhi::ShaderStage::Pixel);
-    
-    auto Vertex = rhi::info<rhi::slotVertex>::make(asset::vertex_i, rhi::VertexInputRate::PerVertex);
+    auto Push_Vertex = rhi::info<rhi::listPush>(rhi::ShaderStage::Vertex,
+      {
+        rhi::vt::make<mat4<f32>>(),
+        rhi::vt::make<u64>(),
+      }
+    );
+
 
-    auto Camera = rhi::info<rhi::slotDesc>::make(ecs::camera_i, rhi::DescriptorType::UniformBuffer, flags(rhi::ShaderStage::Vertex) | rhi::ShaderStage::Pixel);
-    auto Light = rhi::info<rhi::slotDesc>::make(ecs::point_light_i, rhi::DescriptorType::UniformBuffer, flags(rhi::ShaderStage::Vertex) | rhi::ShaderStage::Pixel);
-    auto Texture = rhi::info<rhi::slotDesc>::make(nil, rhi::DescriptorType::TextureSampler, rhi::ShaderStage::Pixel);
+    auto Camera = rhi::info<rhi::slotDesc>(ecs::camera_i, rhi::DescriptorType::UniformBuffer, flags(rhi::ShaderStage::Vertex) | rhi::ShaderStage::Pixel);
+    auto Light = rhi::info<rhi::slotDesc>(ecs::point_light_i, rhi::DescriptorType::UniformBuffer, flags(rhi::ShaderStage::Vertex) | rhi::ShaderStage::Pixel);
+    auto Texture = rhi::info<rhi::slotDesc>(rhi::info<rhi::buffer>(1, {}), rhi::DescriptorType::TextureSampler, rhi::ShaderStage::Pixel);
 
 
-    auto Set0 = rhi::info<rhi::descset>::make({Camera, Light});
-    auto Set1 = rhi::info<rhi::descset>::make({Texture});
+    auto Set0 = rhi::info<rhi::listDesc>({Camera, Light});
+    auto Set1 = rhi::info<rhi::listDesc>({Texture});
 
 
-    shader_color_i = rhi::info<rhi::pipeline>::make({Transform}, {Vertex}, {Set0});
-    shader_texture_i = rhi::info<rhi::pipeline>::make({Transform}, {Vertex}, {Set0, Set1});
+    shader_color_i = rhi::info<rhi::pipeline>({Push_Vertex}, {}, {Set0});
+    shader_texture_i = rhi::info<rhi::pipeline>({Push_Vertex}, {}, {Set0, Set1});
   }
   
 
@@ -156,9 +162,9 @@ namespace mochi::module
       {material_method::mmPBR,  "embed://pbr"},
     };
 
-    std::unordered_map<material_albedo, sptr<rhi::info<rhi::pipeline>>> InAlbedo = {
-      {material_albedo::maColor,   shader_color_i},
-      {material_albedo::maTexture, shader_texture_i},
+    std::unordered_map<material_albedo, rhi::info<rhi::pipeline>*> InAlbedo = {
+      {material_albedo::maColor,   &shader_color_i},
+      {material_albedo::maTexture, &shader_texture_i},
     };
 
     static std::unordered_map<material_albedo, std::string> ShAlbedo = {
@@ -225,7 +231,7 @@ namespace mochi::module
 
     
     auto pipe = rhi::pipeline::make(
-      m_device, InAlbedo[props.albedo], {*vert, *frag},
+      m_device, *InAlbedo[props.albedo], {*vert, *frag},
       props.polymode, props.primitiveTopology,
       static_cast<rhi::Format>(target.color_format), static_cast<rhi::Format>(target.depth_format)
     );
diff --git a/mochi/module/resource.hh b/mochi/module/resource.hh
index eac8722..e866f92 100644
--- a/mochi/module/resource.hh
+++ b/mochi/module/resource.hh
@@ -17,6 +17,7 @@
 #include "mochi/rhi/rhi.hh"
 #include "mochi/types.hh"
 #include "mochi/module/device.hh"
+#include "mochi/rhi/pipeline.hh"
 #include "vulkan/vulkan.hpp"
 #include <expected>
 #include <functional>
@@ -116,7 +117,7 @@ namespace mochi::module
 
 
     private:
-      sptr<rhi::info<rhi::pipeline>> shader_color_i, shader_texture_i;
+      rhi::info<rhi::pipeline> shader_color_i, shader_texture_i;
 
     public:
       fun compile_shader(rhi::ShaderStage stage, std::istream *file, std::vector<std::string> macros) -> std::expected<sptr<rhi::shader>, std::string>;
diff --git a/mochi/res/pbr.vert b/mochi/res/pbr.vert
index 6529e22..0e8b2af 100644
--- a/mochi/res/pbr.vert
+++ b/mochi/res/pbr.vert
@@ -13,12 +13,20 @@
 #version 450
 
 #extension GL_EXT_buffer_reference : require
+#extension GL_EXT_shader_explicit_arithmetic_types_int64 : require
 
 
-layout(location = 0) in vec3 pos;
-layout(location = 1) in vec3 normal;
-layout(location = 2) in vec3 color;
-layout(location = 3) in vec2 uv;
+struct vertex_t {
+  vec3 pos;
+  vec3 normal;
+  vec3 color;
+  vec2 uv;
+};
+
+layout(buffer_reference, std430) readonly buffer VertexBuffer {
+  vertex_t v[];
+};
+
 
 layout(location = 0) out vec3 frag_pos_world;
 layout(location = 1) out vec3 frag_normal_world;
@@ -33,6 +41,7 @@ layout(location = 1) out vec3 frag_normal_world;
 layout(push_constant) uniform PushConstant
 {
   mat4 model;
+  VertexBuffer vertexs;
 } push;
 
 layout(set = 0, binding = 0, row_major) uniform CameraBuffer
@@ -45,19 +54,21 @@ layout(set = 0, binding = 0, row_major) uniform CameraBuffer
 
 void main()
 {
+  vertex_t vex = push.vertexs.v[gl_VertexIndex];
+
   #if defined(WITH_COLOR)
-    frag_color = color;
+    frag_color = vex.color;
   #elif defined(WITH_TEXTURE)
-    frag_uv = uv;
+    frag_uv = vex.uv;
   #endif
 
   mat4 trueModel = transpose(push.model);
 
-  vec4 worldPos = trueModel * vec4(pos, 1.0);
+  vec4 worldPos = trueModel * vec4(vex.pos, 1.0);
   frag_pos_world = worldPos.xyz;
 
   mat3 normalMatrix = transpose(inverse(mat3(trueModel)));
-  frag_normal_world = normalize(normalMatrix * normal);
+  frag_normal_world = normalize(normalMatrix * vex.normal);
 
   gl_Position = camera.proj * camera.view * worldPos;
 }
diff --git a/mochi/rhi/buffer.cc b/mochi/rhi/buffer.cc
index f001634..570a172 100644
--- a/mochi/rhi/buffer.cc
+++ b/mochi/rhi/buffer.cc
@@ -26,10 +26,10 @@
 namespace mochi::rhi
 {
 
-  buffer::buffer(module::device &device, module::memory &memory, sptr<rhi::info<buffer>> info, u64 count, BufferUsageFlags usage, BufferCreateFlags create, BufferLocation location)
+  buffer::buffer(module::device &device, module::memory &memory, rhi::info<buffer> info, u64 count, BufferUsageFlags usage, BufferCreateFlags create, BufferLocation location)
     : m_device(device)
     , m_info(info)
-    , m_size(info->stride() * count)
+    , m_size(info.stride() * count)
     , m_buffer(nil)
     , m_allocator(memory.allocator())
   {
@@ -69,7 +69,7 @@ namespace mochi::rhi
 
   
 
-  fun buffer::make(module::device &device, module::memory &memory, sptr<rhi::info<buffer>> info, u64 count, BufferUsageFlags usage, BufferCreateFlags create, BufferLocation location, std::function<void (void*)> data) -> sptr<buffer>
+  fun buffer::make(module::device &device, module::memory &memory, rhi::info<buffer> info, u64 count, BufferUsageFlags usage, BufferCreateFlags create, BufferLocation location, std::function<void (void*)> data) -> sptr<buffer>
   {
     auto ret = make_sptr(new buffer(
       device, memory,
diff --git a/mochi/rhi/buffer.hh b/mochi/rhi/buffer.hh
index ddef709..f5db6b6 100644
--- a/mochi/rhi/buffer.hh
+++ b/mochi/rhi/buffer.hh
@@ -26,13 +26,8 @@ namespace mochi::rhi
   template<>
   struct info<buffer>
   {
-    private:
-      explicit inline info<buffer>(u64 stride, std::vector<vt> items): m_stride(stride), m_items(items) {}
-
     public:
-      static inline fun make(u64 stride, std::vector<vt> items) {
-        return make_sptr(new info<buffer>(stride, items));
-      }
+      explicit inline info<buffer>(u64 stride, std::vector<vt> items): m_stride(stride), m_items(items) {}
 
 
     private:
@@ -40,8 +35,8 @@ namespace mochi::rhi
       u64 m_stride;
 
     public:
-      inline fun items() { return m_items; }
-      inline fun stride() { return m_stride; }
+      inline fun& items() const { return m_items; }
+      inline fun  stride() const { return m_stride; }
   };
 
 
@@ -50,17 +45,17 @@ namespace mochi::rhi
   struct buffer
   {
     private:
-      explicit buffer(module::device &device, module::memory &memory, sptr<info<buffer>> info, u64 count, BufferUsageFlags usage, BufferCreateFlags create, BufferLocation location);
+      explicit buffer(module::device &device, module::memory &memory, info<buffer> info, u64 count, BufferUsageFlags usage, BufferCreateFlags create, BufferLocation location);
       
     public:
       ~buffer();
       
-      static fun make(module::device &device, module::memory &memory, sptr<info<buffer>> info, u64 count, BufferUsageFlags usage, BufferCreateFlags create, BufferLocation location, std::function<void (void*)> data = nil) -> sptr<buffer>;
+      static fun make(module::device &device, module::memory &memory, info<buffer> info, u64 count, BufferUsageFlags usage, BufferCreateFlags create, BufferLocation location, std::function<void (void*)> data = nil) -> sptr<buffer>;
     
 
-    protected:
+    private:
+      info<buffer> m_info;
       module::device &m_device;
-      sptr<info<buffer>> m_info{};
       u64   m_size{};
       void* m_mapped{};
       VmaAllocator m_allocator{nil};
@@ -69,7 +64,7 @@ namespace mochi::rhi
       VmaAllocationInfo m_alloc_info{};
 
     public:
-      inline fun info() { return m_info.get(); }
+      inline fun info() { return &m_info; }
       inline fun size() const { return m_size; }
       inline fun mapped() const { return m_mapped; }
       inline fun get() const { return m_buffer; }
diff --git a/mochi/rhi/convert.hh b/mochi/rhi/convert.hh
index e8c1586..cdcd0e1 100644
--- a/mochi/rhi/convert.hh
+++ b/mochi/rhi/convert.hh
@@ -114,6 +114,14 @@ namespace mochi::rhi
   }
 
 
+  template <typename T>
+    requires (std::is_same_v<T, PipelineKind>)
+  inline fun VKConvert(PipelineKind in) -> vk::PipelineBindPoint
+  {
+    return static_cast<vk::PipelineBindPoint>(in);
+  }
+
+
   template <typename T>
     requires (std::is_same_v<T, Format>)
   inline fun VKConvert(Format in) -> vk::Format
diff --git a/mochi/rhi/descset.hh b/mochi/rhi/descset.hh
deleted file mode 100644
index 05a3287..0000000
--- a/mochi/rhi/descset.hh
+++ /dev/null
@@ -1,60 +0,0 @@
-/*
-  This file is part of QAOS
-
-  This file is licensed under the GNU General Public License version 3 (GPL3).
-
-  You should have received a copy of the GNU General Public License
-  along with QAOS. If not, see <https://www.gnu.org/licenses/>.
-
-  Copyright (c) 2025-2026 by Kadir Aydın.
-*/
-
-
-#pragma once
-
-#include "mochi/basis.hh"
-#include "mochi/types.hh"
-#include <cassert>
-
-
-
-namespace mochi::rhi
-{
-
-  template<>
-  struct info<descset>
-  {
-    private:
-      info(std::vector<sptr<info<slotDesc>>> idescs): m_idescs(idescs) {}
-
-    public:
-      static inline fun make(std::vector<sptr<info<slotDesc>>> idescs) {
-        return make_sptr(new info<descset>(idescs));
-      }
-
-
-    private:
-      std::vector<sptr<info<slotDesc>>> m_idescs;
-
-    public:
-      inline fun& idescs() { return m_idescs; }
-  };
-
-  
-
-  struct descset
-  {
-    public:
-      descset(u32 id, std::vector<sptr<slotDesc>> descs): m_id(id), m_descs(descs) {}
-
-
-    private:
-      u32 m_id;
-      std::vector<sptr<slotDesc>> m_descs;
-    
-    public:
-      inline fun id() { return m_id; }
-      inline fun descs() { return m_descs; }
-  };
-
-}
diff --git a/mochi/rhi/image.cc b/mochi/rhi/image.cc
index a46190d..61bc0bf 100644
--- a/mochi/rhi/image.cc
+++ b/mochi/rhi/image.cc
@@ -31,7 +31,7 @@ namespace mochi::rhi
   {
     vk::DeviceSize image_size = width * height * 4;
 
-    auto staging_info = rhi::info<rhi::buffer>::make(1, {}); 
+    auto staging_info = rhi::info<rhi::buffer>(1, {}); 
     
     VmaAllocationCreateInfo staging_alloc_info = {};
     staging_alloc_info.usage = VMA_MEMORY_USAGE_AUTO;
diff --git a/mochi/rhi/image.hh b/mochi/rhi/image.hh
index b8c0a57..ffdb401 100644
--- a/mochi/rhi/image.hh
+++ b/mochi/rhi/image.hh
@@ -35,7 +35,7 @@ namespace mochi::rhi
       }
 
 
-    protected:
+    private:
       u32 m_width, m_height;
       VmaAllocator m_allocator{nil};
       VkImage m_image{VK_NULL_HANDLE};
diff --git a/mochi/rhi/pipeline.cc b/mochi/rhi/pipeline.cc
index 33da5f0..ed7710a 100644
--- a/mochi/rhi/pipeline.cc
+++ b/mochi/rhi/pipeline.cc
@@ -14,9 +14,9 @@
 #include "mochi/rhi/rhi.hh"
 #include "mochi/types.hh"
 #include "mochi/rhi/convert.hh"
-#include "mochi/rhi/buffer.hh"
 #include "mochi/rhi/pipeline.hh"
-#include "mochi/rhi/descset.hh"
+#include "mochi/rhi/listDesc.hh"
+#include "mochi/rhi/listPush.hh"
 #include "mochi/rhi/slotPush.hh"
 #include "mochi/rhi/slotDesc.hh"
 #include "mochi/rhi/slotVertex.hh"
@@ -32,28 +32,32 @@ namespace mochi::rhi
 
   constexpr inline fun align_size(u64 size, u64 alignment) { if (alignment == 0) return size; else return ((size + alignment - 1) / alignment) * alignment; }
 
-  info<pipeline>::info(std::vector<sptr<info<slotPush>>> push, std::vector<sptr<info<slotVertex>>> vertex, std::vector<sptr<info<rhi::descset>>> descset)
+  info<pipeline>::info(std::vector<info<listPush>> pushlist, std::vector<info<slotVertex>> vertex, std::vector<info<rhi::listDesc>> descset)
   {
-    u64 off{};
-    for (auto &typ: push) {
-      off = align_size(off, typ->type().align());
+    u16 off{};
+    for (auto &push: pushlist) {
+      u16 size{};
+
+      auto legoff = off;
+      for (auto &typ: push.ipush()) {
+        off = align_size(off, typ.type().align());
+        off += typ.type().size() * typ.type().count();
+      }
 
       vk_PushConstant.push_back(vk::PushConstantRange(
-        VKConvert<ShaderStageFlags>(typ->shaderStage()),
-        off,
-        typ->type().size() * typ->type().count()
+        VKConvert<ShaderStageFlags>(push.stage()),
+        legoff,
+        off-legoff
       ));
-
-      off += typ->type().size() * typ->type().count();
     }
 
 
 
     u32 binding_idx{}, location_idx{};
-    for (auto ibuf: vertex) {
+    for (auto &ibuf: vertex) {
       u32 v_offset{};
 
-      for (auto &typ: ibuf->ibuf()->items()) {
+      for (auto &typ: ibuf.ibuf()->items()) {
         v_offset = align_size(v_offset, typ.align());
         
         for (u32 c{}; c < typ.count(); c++) {
@@ -69,9 +73,11 @@ namespace mochi::rhi
 
       vk_vertexBinding.push_back(vk::VertexInputBindingDescription(
         binding_idx++,
-        ibuf->ibuf()->stride(),
-        VKConvert<VertexInputRate>(ibuf->inputRate())
-      )); 
+        ibuf.ibuf()->stride(),
+        VKConvert<VertexInputRate>(ibuf.inputRate())
+      ));
+
+      ibuf.index() = binding_idx;
     }
   
     
@@ -83,13 +89,13 @@ namespace mochi::rhi
       std::vector<vk::DescriptorSetLayoutBinding> current_set_bindings;
       
       u32 binding_idx{};
-      for (auto &slot: desc->idescs()) 
+      for (auto &slot: desc.idescs()) 
       {
         current_set_bindings.push_back(vk::DescriptorSetLayoutBinding(
           binding_idx++,
-          VKConvert<DescriptorType>(slot->kind()),
+          VKConvert<DescriptorType>(slot.kind()),
           1,
-          VKConvert<ShaderStageFlags>(slot->stage()),
+          VKConvert<ShaderStageFlags>(slot.stage()),
           nil
         ));
       }
@@ -98,18 +104,19 @@ namespace mochi::rhi
     }
 
     
-    m_push    = std::move(push);
+    m_push    = std::move(pushlist);
     m_vertex  = std::move(vertex);
     m_descset = std::move(descset);
   }
 
 
   pipeline::pipeline(
-    module::device &device, sptr<rhi::info<pipeline>> info, std::vector<sptr<shader>> shaders,
+    module::device &device, rhi::info<pipeline> info, std::vector<sptr<shader>> shaders,
     PolygonMode polymode, PrimitiveTopology primitiveTopology,
     Format __color_format, Format __depth_format
   )
     : vk_layout(nil), vk_pipeline(nil)
+    , m_dbg_info(info)
   {
     bool is_compute = (shaders.size() == 1 && shaders[0]->stage() == ShaderStage::Compute);
 
@@ -117,7 +124,7 @@ namespace mochi::rhi
     std::vector<vk::DescriptorSetLayout> raw_layouts;
 
 
-    for (auto &bindings_for_set: info->descBindings()) {
+    for (auto &bindings_for_set: info.descBindings()) {
       vk::DescriptorSetLayoutCreateInfo set_info({}, bindings_for_set);
       
       vk_desc_layouts.push_back(vk::raii::DescriptorSetLayout(device.vdevice(), set_info));
@@ -126,13 +133,14 @@ namespace mochi::rhi
     }
     
 
-    vk::PipelineLayoutCreateInfo layout_info({}, raw_layouts, info->pushConstant());
+    vk::PipelineLayoutCreateInfo layout_info({}, raw_layouts, info.pushConstant());
     vk_layout = vk::raii::PipelineLayout(device.vdevice(), layout_info);
 
 
 
     if (is_compute) 
     {
+      m_kind = PipelineKind::Compute;
       vk::PipelineShaderStageCreateInfo compute_stage(
         {}, VKConvert<ShaderStage>(shaders[0]->stage()), shaders[0]->module(), shaders[0]->entry().c_str()
       );
@@ -144,6 +152,7 @@ namespace mochi::rhi
     } 
     else 
     {
+      m_kind = PipelineKind::Graphic;
       std::vector<vk::PipelineShaderStageCreateInfo> shader_stages;
       shader_stages.reserve(shaders.size());
       for (auto &sh: shaders) {
@@ -184,7 +193,7 @@ namespace mochi::rhi
         depth_format, vk::Format::eUndefined
       );
 
-      auto VertexInput = info->vertexBinding();
+      auto VertexInput = info.vertexBinding();
       vk::GraphicsPipelineCreateInfo pipeline_info(
         {}, shader_stages, &VertexInput, &input_assembly, nil,
         &viewport_state, &rasterizer, &multisampling, &depth_stencil,
diff --git a/mochi/rhi/pipeline.hh b/mochi/rhi/pipeline.hh
index 1af8ed0..dfdae0c 100644
--- a/mochi/rhi/pipeline.hh
+++ b/mochi/rhi/pipeline.hh
@@ -16,7 +16,8 @@
 #include "mochi/rhi/rhi.hh"
 #include "mochi/types.hh"
 #include "mochi/rhi/shader.hh"
-#include "mochi/rhi/slotPush.hh"
+#include "mochi/rhi/listPush.hh"
+#include "mochi/rhi/listDesc.hh"
 #include "mochi/rhi/slotVertex.hh"
 #include <vulkan/vulkan.hpp>
 #include <vulkan/vulkan_raii.hpp>
@@ -29,25 +30,15 @@ namespace mochi::rhi
   template<>
   struct info<pipeline>
   {
-    private:
-      explicit info(std::vector<sptr<info<slotPush>>> push, std::vector<sptr<info<slotVertex>>> vertex, std::vector<sptr<info<descset>>> descset);
-
-    public:
-      static inline fun make(std::vector<sptr<info<slotPush>>> push, std::vector<sptr<info<slotVertex>>> vertex, std::vector<sptr<info<descset>>> descset) {
-        return make_sptr(new info<pipeline>(std::move(push), std::move(vertex), std::move(descset)));
-      }
-
-
-    protected:
-      info() = default;
     public:
-      virtual ~info() = default;
+      explicit inline info(nil_t): m_push() {}
+      explicit info(std::vector<info<listPush>> pushlist, std::vector<info<slotVertex>> vertex, std::vector<info<listDesc>> descset);
 
 
     protected:
-      std::vector<sptr<info<slotPush>>>   m_push;
-      std::vector<sptr<info<slotVertex>>> m_vertex;
-      std::vector<sptr<info<descset>>>    m_descset;
+      std::vector<info<listPush>>   m_push;
+      std::vector<info<slotVertex>> m_vertex;
+      std::vector<info<listDesc>>    m_descset;
 
       std::vector<vk::PushConstantRange> vk_PushConstant;
       std::vector<vk::VertexInputBindingDescription> vk_vertexBinding;
@@ -56,9 +47,9 @@ namespace mochi::rhi
 
 
     public:
-      inline const fun& push() { return m_push; }
-      inline const fun& vertex() { return m_vertex; }
-      inline const fun& descset() { return m_descset; }
+      inline fun& push() { return m_push; }
+      inline fun& vertex() { return m_vertex; }
+      inline fun& desc() { return m_descset; }
 
       inline fun& pushConstant() { return vk_PushConstant; }
       inline fun  vertexBinding() { return vk::PipelineVertexInputStateCreateInfo({}, vk_vertexBinding, vk_vertexAttribute); }
@@ -72,14 +63,14 @@ namespace mochi::rhi
   {
     private:
       explicit pipeline(
-        module::device &device, sptr<info<pipeline>> info, std::vector<sptr<shader>> shaders,
+        module::device &device, info<pipeline> info, std::vector<sptr<shader>> shaders,
         PolygonMode polymode, PrimitiveTopology primitiveTopology,
         Format color_format, Format depth_format
       );
       
     public:
       static inline fun make(
-        module::device &device, sptr<info<pipeline>> info, std::vector<sptr<shader>> shaders,
+        module::device &device, info<pipeline> info, std::vector<sptr<shader>> shaders,
         PolygonMode polymode, PrimitiveTopology primitiveTopology,
         Format color_format, Format depth_format
       ) {
@@ -88,14 +79,17 @@ namespace mochi::rhi
     
 
     private:
-      sptr<info<pipeline>>     m_info{};
+      info<pipeline> m_dbg_info;
+
+      PipelineKind m_kind;
       vk::raii::PipelineLayout vk_layout;
       vk::raii::Pipeline       vk_pipeline;
       std::vector<vk::raii::DescriptorSetLayout> vk_desc_layouts;
       
     public:
-      inline fun  info() { return m_info.get(); }
+      inline fun& info() { return m_dbg_info; }
       inline fun& get() { return vk_pipeline; }
+      inline fun  kind() { return m_kind; }
       inline fun& layout() { return vk_layout; }
       inline fun& desc_layouts() { return vk_desc_layouts; }
   };
diff --git a/mochi/rhi/rhi.hh b/mochi/rhi/rhi.hh
index 78e0492..c027fb2 100644
--- a/mochi/rhi/rhi.hh
+++ b/mochi/rhi/rhi.hh
@@ -107,6 +107,14 @@ namespace mochi::rhi
   };
 
 
+  enum struct PipelineKind: u32
+  {
+    Graphic    = 0,
+    Compute    = 1,
+    RayTracing = 1000165000,
+  };
+
+
   enum struct Format: u32
   {
     /// 8-bit
diff --git a/mochi/rhi/slotDesc.hh b/mochi/rhi/slotDesc.hh
index a0b9162..0d55735 100644
--- a/mochi/rhi/slotDesc.hh
+++ b/mochi/rhi/slotDesc.hh
@@ -14,6 +14,7 @@
 
 #include "mochi/basis.hh"
 #include "mochi/rhi/rhi.hh"
+#include "mochi/rhi/buffer.hh"
 #include "mochi/types.hh"
 #include <cassert>
 
@@ -25,26 +26,21 @@ namespace mochi::rhi
   template<>
   struct info<slotDesc>
   {
-    private:
-      info(sptr<info<buffer>> ibuf, DescriptorType kind, ShaderStageFlags stage)
+    public:
+      info(info<buffer> ibuf, DescriptorType kind, ShaderStageFlags stage)
         : m_ibuf(ibuf)
         , m_kind(kind)
         , m_stage(stage)
       {}
 
-    public:
-      static inline fun make(sptr<info<buffer>> ibuf, DescriptorType kind, ShaderStageFlags stage) {
-        return make_sptr(new info<slotDesc>(ibuf, kind, stage));
-      }
-
 
     private:
-      sptr<info<buffer>> m_ibuf;
+      info<buffer> m_ibuf;
       DescriptorType m_kind;
       ShaderStageFlags m_stage;
 
     public:
-      inline fun ibuf() { return m_ibuf.get(); }
+      inline fun ibuf() { return &m_ibuf; }
       inline fun kind() { return m_kind; }
       inline fun stage() { return m_stage; }
   };
@@ -54,21 +50,21 @@ namespace mochi::rhi
   struct slotDesc
   {
     public:
-      slotDesc(info<slotDesc> *info, rhi::buffer *buf)
+      slotDesc(info<slotDesc> info, rhi::buffer *buf)
         : m_info(info)
         , m_buf(buf)
       {
-        assert(info && buf);
+        assert(buf);
       }
 
 
     private:
-      info<slotDesc> *m_info;
+      info<slotDesc> m_info;
       rhi::buffer *m_buf;
     
     public:
-      inline fun info() { return m_info; }
-      inline fun buf() { return m_buf; }
+      inline fun info() const { return &m_info; }
+      inline fun buf() const { return m_buf; }
   };
 
 }
diff --git a/mochi/rhi/slotPush.hh b/mochi/rhi/slotPush.hh
index 0fc9d3f..22fc23b 100644
--- a/mochi/rhi/slotPush.hh
+++ b/mochi/rhi/slotPush.hh
@@ -13,10 +13,11 @@
 #pragma once
 
 #include "mochi/basis.hh"
-#include "mochi/rhi/rhi.hh"
 #include "mochi/rhi/vtype.hh"
 #include "mochi/types.hh"
 #include <cassert>
+#include <cstdlib>
+#include <iostream>
 
 
 
@@ -26,47 +27,51 @@ namespace mochi::rhi
   template<>
   struct info<slotPush>
   {
-    private:
-      info(vt type, ShaderStageFlags shaderStage)
-        : m_type(type)
-        , m_shaderStage(shaderStage)
-      {}
-
     public:
-      static inline fun make(vt type, ShaderStageFlags shaderStage) {
-        return make_sptr(new info<slotPush>(type, shaderStage));
-      }
+      info(vt type): m_type(type) {}
 
 
     private:
       vt m_type;
-      ShaderStageFlags m_shaderStage;
+      u32 m_offset{};
 
     public:
-      inline fun type() { return m_type; }
-      inline fun shaderStage() { return m_shaderStage; }
+      inline fun  type() const { return m_type; }
+      inline fun& offset() { return m_offset; }
   };
 
 
 
+
   struct slotPush
   {
     public:
-      slotPush(info<slotPush> *info, rhi::buffer *buf)
-        : m_info(info)
-        , m_buf(buf)
+      template <typename T>
+      inline slotPush(T raw)
+        : m_data(
+          std::malloc(sizeof(T)),
+          sizeof(T)
+        )
       {
-        assert(info && buf);
+        T vaw = raw;
+        __builtin_memcpy(m_data.ptr(), &vaw, sizeof(T));
+        std::cerr << "malloc: " << m_data.ptr() << std::endl;
       }
 
+      inline ~slotPush() {
+        std::cerr << "free: " << m_data.ptr() << std::endl;
+        std::free(m_data.ptr());
+      }
+
+      slotPush(const slotPush&) = delete;
+      slotPush& operator=(const slotPush&) = delete;
+
 
     private:
-      info<slotPush> *m_info;
-      rhi::buffer *m_buf;
-    
+      data m_data{};
+      
     public:
-      inline fun info() { return m_info; }
-      inline fun buf() { return m_buf; }
+      inline fun data() const { return m_data; }
   };
-
+ 
 }
diff --git a/mochi/rhi/slotVertex.hh b/mochi/rhi/slotVertex.hh
index a065228..026fa91 100644
--- a/mochi/rhi/slotVertex.hh
+++ b/mochi/rhi/slotVertex.hh
@@ -14,6 +14,7 @@
 
 #include "mochi/basis.hh"
 #include "mochi/rhi/rhi.hh"
+#include "mochi/rhi/buffer.hh"
 #include "mochi/types.hh"
 #include <cassert>
 
@@ -25,25 +26,23 @@ namespace mochi::rhi
   template<>
   struct info<slotVertex>
   {
-    private:
-      info(sptr<info<rhi::buffer>> ibuf, VertexInputRate inputRate)
+    public:
+      info(info<rhi::buffer> ibuf, VertexInputRate inputRate, i16 index = -1)
         : m_ibuf(ibuf)
         , m_inputRate(inputRate)
+        , m_index(index)
       {}
 
-    public:
-      static inline fun make(sptr<info<rhi::buffer>> ibuf, VertexInputRate inputRate) {
-        return make_sptr(new info<slotVertex>(ibuf, inputRate));
-      }
-
 
     private:
-      sptr<info<rhi::buffer>> m_ibuf;
+      info<rhi::buffer> m_ibuf;
+      i16 m_index;
       VertexInputRate m_inputRate;
 
     public:
-      inline fun ibuf() { return m_ibuf.get(); }
-      inline fun inputRate() { return m_inputRate; }
+      inline fun  ibuf() const { return &m_ibuf; }
+      inline fun  inputRate() const { return m_inputRate; }
+      inline fun& index() { return m_index; }
   };
 
 
@@ -51,21 +50,24 @@ namespace mochi::rhi
   struct slotVertex
   {
     public:
-      slotVertex(info<slotVertex> *info, rhi::buffer *buf)
+      slotVertex(info<slotVertex> info, rhi::buffer *buf, i16 index = -1)
         : m_info(info)
         , m_buf(buf)
+        , m_index(index)
       {
-        assert(info && buf);
+        assert(buf);
       }
 
 
     private:
-      info<slotVertex> *m_info;
+      info<slotVertex> m_info;
+      i16 m_index;
       rhi::buffer *m_buf;
     
     public:
-      inline fun info() { return m_info; }
-      inline fun buf() { return m_buf; }
+      inline fun index() const { return m_index; }
+      inline fun info() const { return &m_info; }
+      inline fun buf() const { return m_buf; }
   };
 
 }
diff --git a/mochi/rhi/vtype.hh b/mochi/rhi/vtype.hh
index ed21114..f169e75 100644
--- a/mochi/rhi/vtype.hh
+++ b/mochi/rhi/vtype.hh
@@ -162,10 +162,10 @@ namespace mochi::rhi
       u8 m_count;
 
     public:
-      inline fun size() { return m_size; }
-      inline fun align() { return m_align; }
-      inline fun format() { return m_format; }
-      inline fun count() { return m_count; }
+      inline fun size() const { return m_size; }
+      inline fun align() const { return m_align; }
+      inline fun format() const { return m_format; }
+      inline fun count() const { return m_count; }
   };
 
 }
diff --git a/mochi/types.hh b/mochi/types.hh
index 8b88aee..b215484 100644
--- a/mochi/types.hh
+++ b/mochi/types.hh
@@ -39,7 +39,9 @@ namespace mochi
     struct slotPush;
     struct slotDesc;
     struct slotVertex;
-    struct descset;
+    
+    struct listDesc;
+    struct listPush;
 
     struct pipeline;
     struct buffer;
